spyql/parser.py

```diff
@@ -86,42 +86,40 @@ def parse_structure(query: str):
     query_struct: Dict[Optional[str], Optional[str]] = {
         kw: None for kw in query_struct_keywords
     }
+
+    # {"select": ["select"], ..., "group": ["group", "by"], ...}
+    query_struct_keyword_dict: Dict[str, List[str]] = {
+        mk[0]: mk for mk in [k.split() for k in query_struct_keywords]
+    }
     present_keyword = None
 
     validator = KeywordOrderValidator(query_struct_keywords)
 
     i = 0
     while i < len(tokens):
-        token = tokens[i].lower()
-        next_token = tokens[i + 1].lower() if i < len(tokens) - 1 else ""
+        token = tokens[i]
 
-        # For single keyword
-        if token in query_struct_keywords:
-            present_keyword = token
+        keyword_parts = query_struct_keyword_dict.get(token.lower())
+        if keyword_parts and keyword_parts == [
+            t.lower() for t in tokens[i : i + len(keyword_parts)]
+        ]:
+            present_keyword = " ".join(keyword_parts)
             validator.run(present_keyword)
-
             query_struct[present_keyword] = ""
-            i += 1
-        # For multiple-word keyword
-        elif i < len(tokens) - 1 and token + " " + next_token in query_struct_keywords:
-            present_keyword = token + " " + next_token
-            validator.run(present_keyword)
-            query_struct[present_keyword] = ""
-            i += 2
+            i += len(keyword_parts)
+
         # For not-keyword token
         else:
-            # For not-keyword token, `token` variable is not always valid
-            # because `token` is the "lower case" of tokens[i]
             if not present_keyword:
                 log.user_error(
                     "could not parse query",
-                    SyntaxError(f"misplaced '{tokens[i]}' clause"),
+                    SyntaxError(f"misplaced '{token}' clause"),
                 )
 
             if query_struct[present_keyword]:
-                query_struct[present_keyword] += f" {tokens[i]}"
+                query_struct[present_keyword] += f" {token}"
             else:
-                query_struct[present_keyword] += tokens[i]
+                query_struct[present_keyword] += token
```

Comment:
That is almost the same logic as before the change. And, that method seems inefficient as it requires looking at the query once to find the keyword position and once more to extract the tokens.

Reply:
I understand where you are coming from, but the reasoning when programming in an interpreted language like Python is very different from when programming in compiled languages like C. The number of operations is not a good metric because, for instance, a list comprehension is much faster than a for loop. In the same way, a string.join of a list of strings is much faster than using a for loop to go through the list and concatenating as we go. This happens because 1) there is an overhead for each line of code that is called and 2) python makes available very efficient implementations for common tasks like joining multiple strings. 

I think that most of the gains would come from what I called the 2nd level optimization. The only way to be sure is to implement and measure each alternative and then assess performance gains vs code readability and elegance. 

You don't need to implement any of those: your contribution, as it is, is of high quality and solves the issues that it addresses.

Just let me know how you would like to proceed.

Reply:
Your third level optimization is shortest and fastest.

```python
def third(tokens: List[str]):
    candidates = [
        (i, query_struct_keyword_dict[token.lower()])
        for i, token in enumerate(tokens)
        if token.lower() in query_struct_keyword_dict
    ]

    for i, (pos, kw) in enumerate(candidates):
        if i < len(candidates) - 1:
            pos_next = candidates[i + 1][0]
            query_struct[" ".join(kw)] = " ".join(tokens[pos + len(kw) : pos_next])
        else:
            query_struct[" ".join(kw)] = " ".join(tokens[pos + len(kw) :])

    return query_struct

if __name__ == "__main__":
    query = "SELECT * FROM range(0, 10) ORDER BY 1 TO json"
    tokens: List[str] = []
    for line in query.splitlines():
        if "#" in line:
            # Remove comment from parse targets
            line = line.split("#", 1)[0]

        tokens += line.split()

    tokens = [t for t in tokens if t != ""]

    query_struct: Dict[Optional[str], Optional[str]] = {
        kw: None for kw in query_struct_keywords
    }

    # {"select": ["select"], ..., "group": ["group", "by"], ...}
    query_struct_keyword_dict: Dict[str, List[str]] = {
        mk[0]: mk for mk in [k.split() for k in query_struct_keywords]
    }

    for i in range(1000000):
        _ = third(tokens)
```

Third optimization: 2.65 sec

Reply:
Thank you for trying @Hayashi-Yudai 😄  
That's the spirit! 💪 

I must say that I did an implementation of the 2nd optimisation and saw improvements on my machine. But, these should depend on the size of the query (number of spaces and keywords) and may vary for different hardware/OS. 

Here's my implementation, which still needs a minor improvement:

```python
   present_keyword = None               
   i = 0
   start_idx = None
   while i < len(tokens):
        token = tokens[i]
        
        keyword_parts = query_struct_keyword_dict.get(token.lower())
        if keyword_parts and keyword_parts == [
            t.lower() for t in tokens[i : i + len(keyword_parts)]
        ]:
            if present_keyword:
              query_struct[present_keyword] = " ".join(tokens[start_idx:i])
            present_keyword = " ".join(keyword_parts)
            validator.run(present_keyword)
            i += len(keyword_parts)
            start_idx = i 
        
        # For not-keyword token
        else:
            if not present_keyword: # this check could be moved elsewhere
                log.user_error(
                    "could not parse query",
                    SyntaxError(f"misplaced '{token}' clause"),
                )
            i += 1
            
    if present_keyword:  
        # when we get to the end of tokens, we need to close the current keyword 
        query_struct[present_keyword] = " ".join(tokens[start_idx:])
    return query_struct
```

Reply:
Were's an improved 2nd, without a while loop:

Reply:
@Hayashi-Yudai: you really nailed it in the third implementation 👏